File: pcces-go/internal/storage/sqlite/cost_structure_calculation.go
package sqlite

import (
	"fmt"
	"sort"
	"strings"

	"github.com/shopspring/decimal"
)

type CostCalculationLine struct {
	Code        string `json:"code"`
	Kind        string `json:"kind"`
	BaseKind    string `json:"base_kind"`
	Rate        string `json:"rate"`
	FixedAmount string `json:"fixed_amount"`
	Sign        int    `json:"sign"`
	SortOrder   int    `json:"sort_order"`
}

type CostCalculationResultLine struct {
	Code         string `json:"code"`
	Kind         string `json:"kind"`
	BaseKind     string `json:"base_kind"`
	BaseAmount   string `json:"base_amount"`
	Rate         string `json:"rate"`
	Sign         int    `json:"sign"`
	Amount       string `json:"amount"`
	RunningTotal string `json:"running_total"`
	SortOrder    int    `json:"sort_order"`
}

type CostCalculationResult struct {
	DirectCost string                      `json:"direct_cost"`
	Total      string                      `json:"total"`
	Scale      int32                       `json:"scale"`
	Lines      []CostCalculationResultLine `json:"lines"`
	Trace      map[string]any              `json:"calculation_trace"`
}

func CalculateCostStructure(lines []CostCalculationLine, directCost string, scale int32) (CostCalculationResult, error) {
	if scale < 0 || scale > 8 {
		return CostCalculationResult{}, fmt.Errorf("scale must be between 0 and 8")
	}
	direct, err := decimal.NewFromString(directCost)
	if err != nil {
		return CostCalculationResult{}, fmt.Errorf("invalid direct_cost: %w", err)
	}
	sort.SliceStable(lines, func(i, j int) bool {
		if lines[i].SortOrder == lines[j].SortOrder {
			return strings.ToUpper(lines[i].Code) < strings.ToUpper(lines[j].Code)
		}
		return lines[i].SortOrder < lines[j].SortOrder
	})
	allowedKinds := map[string]bool{"DIRECT": true, "INDIRECT": true, "MANAGEMENT": true, "TAX": true, "PERCENT": true, "ADJUSTMENT": true}
	allowedBases := map[string]bool{"DIRECT": true, "SUBTOTAL": true, "PREVIOUS": true, "FIXED": true}
	seen := map[string]bool{}
	subtotal := direct.Round(scale)
	previous := decimal.Zero
	result := CostCalculationResult{DirectCost: direct.Round(scale).StringFixed(scale), Scale: scale, Lines: []CostCalculationResultLine{}}
	order := []string{}
	for _, raw := range lines {
		code := strings.ToUpper(strings.TrimSpace(raw.Code))
		kind := strings.ToUpper(strings.TrimSpace(raw.Kind))
		baseKind := strings.ToUpper(strings.TrimSpace(raw.BaseKind))
		if baseKind == "" {
			baseKind = "SUBTOTAL"
		}
		if code == "" || seen[code] {
			return CostCalculationResult{}, fmt.Errorf("line code is required and must be unique")
		}
		if !allowedKinds[kind] || !allowedBases[baseKind] {
			return CostCalculationResult{}, fmt.Errorf("unsupported kind or base_kind")
		}
		if raw.Sign != -1 && raw.Sign != 1 {
			return CostCalculationResult{}, fmt.Errorf("sign must be -1 or 1")
		}
		seen[code] = true
		rate, err := decimal.NewFromString(defaultString(raw.Rate, "0"))
		if err != nil {
			return CostCalculationResult{}, fmt.Errorf("invalid rate: %w", err)
		}
		fixed, err := decimal.NewFromString(defaultString(raw.FixedAmount, "0"))
		if err != nil {
			return CostCalculationResult{}, fmt.Errorf("invalid fixed_amount: %w", err)
		}
		base := subtotal
		switch baseKind {
		case "DIRECT":
			base = direct
		case "PREVIOUS":
			base = previous
		case "FIXED":
			base = fixed
		}
		amount := base.Mul(rate).Div(decimal.NewFromInt(100))
		if kind == "ADJUSTMENT" || baseKind == "FIXED" {
			amount = fixed
		}
		amount = amount.Mul(decimal.NewFromInt(int64(raw.Sign))).Round(scale)
		subtotal = subtotal.Add(amount).Round(scale)
		previous = amount
		result.Lines = append(result.Lines, CostCalculationResultLine{Code: code, Kind: kind, BaseKind: baseKind, BaseAmount: base.Round(scale).StringFixed(scale), Rate: rate.String(), Sign: raw.Sign, Amount: amount.StringFixed(scale), RunningTotal: subtotal.StringFixed(scale), SortOrder: raw.SortOrder})
		order = append(order, code)
	}
	result.Total = subtotal.StringFixed(scale)
	result.Trace = map[string]any{"policy": "P4-COST-005", "rounding": "ROUND_HALF_UP", "order": order}
	return result, nil
}

func defaultString(value, fallback string) string {
	if strings.TrimSpace(value) == "" {
		return fallback
	}
	return value
}

File: api/cost_structure_calculation.py
"""Phase 4 cost-structure initialization and deterministic fee calculation."""
from __future__ import annotations

from decimal import Decimal, ROUND_HALF_UP
from flask import Blueprint, jsonify, request

ALLOWED_KINDS = {"DIRECT", "INDIRECT", "MANAGEMENT", "TAX", "PERCENT", "ADJUSTMENT"}
ALLOWED_BASES = {"DIRECT", "SUBTOTAL", "PREVIOUS", "FIXED"}


def q(value: Decimal, scale: int) -> Decimal:
    unit = Decimal(1).scaleb(-scale)
    return value.quantize(unit, rounding=ROUND_HALF_UP)


def calculate_cost_structure(lines: list[dict], direct_cost: str, scale: int = 2) -> dict:
    if not 0 <= scale <= 8:
        raise ValueError("scale must be between 0 and 8")
    direct = Decimal(str(direct_cost))
    ordered = sorted(lines, key=lambda item: (int(item.get("sort_order", 0)), str(item.get("code", ""))))
    seen: set[str] = set()
    subtotal = direct
    previous = Decimal("0")
    results: list[dict] = []
    for raw in ordered:
        code = str(raw.get("code", "")).strip().upper()
        kind = str(raw.get("kind", "")).strip().upper()
        base_kind = str(raw.get("base_kind", "SUBTOTAL")).strip().upper()
        if not code or code in seen:
            raise ValueError("line code is required and must be unique")
        if kind not in ALLOWED_KINDS:
            raise ValueError(f"unsupported kind: {kind}")
        if base_kind not in ALLOWED_BASES:
            raise ValueError(f"unsupported base_kind: {base_kind}")
        seen.add(code)
        sign = int(raw.get("sign", 1))
        if sign not in {-1, 1}:
            raise ValueError("sign must be -1 or 1")
        rate = Decimal(str(raw.get("rate", "0")))
        fixed = Decimal(str(raw.get("fixed_amount", "0")))
        if base_kind == "DIRECT":
            base = direct
        elif base_kind == "PREVIOUS":
            base = previous
        elif base_kind == "FIXED":
            base = fixed
        else:
            base = subtotal
        amount = fixed if kind == "ADJUSTMENT" or base_kind == "FIXED" else base * rate / Decimal("100")
        amount = q(amount * sign, scale)
        subtotal = q(subtotal + amount, scale)
        previous = amount
        results.append({
            "code": code, "kind": kind, "base_kind": base_kind,
            "base_amount": format(q(base, scale), "f"),
            "rate": format(rate, "f"), "sign": sign,
            "amount": format(amount, "f"), "running_total": format(subtotal, "f"),
            "sort_order": int(raw.get("sort_order", 0)),
        })
    return {
        "direct_cost": format(q(direct, scale), "f"),
        "total": format(subtotal, "f"), "scale": scale,
        "lines": results,
        "calculation_trace": {
            "policy": "P4-COST-005",
            "rounding": "ROUND_HALF_UP",
            "order": [item["code"] for item in results],
        },
    }


def build_cost_structure_calculation_blueprint(resolve_user_id):
    bp = Blueprint("cost_structure_calculation", __name__, url_prefix="/api/cost-structures")

    @bp.post("/calculate")
    def calculate():
        if resolve_user_id() is None:
            return jsonify({"code": "UNAUTHORIZED"}), 401
        body = request.get_json(silent=True) or {}
        try:
            return jsonify(calculate_cost_structure(
                list(body.get("lines") or []), str(body.get("direct_cost", "0")), int(body.get("scale", 2))
            ))
        except (ValueError, ArithmeticError) as exc:
            return jsonify({"code": "INVALID_ARGUMENT", "detail": str(exc)}), 400

    return bp
